Sources/AsyncHTTPClient/HTTPClient.swift

```diff
@@ -64,24 +68,80 @@ public class HTTPClient {
             self.eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
         }
         self.configuration = configuration
+        self.pool = ConnectionPool(configuration: configuration)
+        self.state = .upAndRunning
     }
 
     deinit {
-        assert(self.isShutdown.load(), "Client not shut down before the deinit. Please call client.syncShutdown() when no longer needed.")
+        assert(self.state == .shutDown, "Client not shut down before the deinit. Please call client.syncShutdown() when no longer needed.")
     }
 
     /// Shuts down the client and `EventLoopGroup` if it was created by the client.
     public func syncShutdown() throws {
-        switch self.eventLoopGroupProvider {
-        case .shared:
-            self.isShutdown.store(true)
-            return
-        case .createNew:
-            if self.isShutdown.compareAndExchange(expected: false, desired: true) {
-                try self.eventLoopGroup.syncShutdownGracefully()
-            } else {
+        try self.syncShutdown(requiresCleanClose: false)
+    }
+
+    /// Shuts down the client and `EventLoopGroup` if it was created by the client.
+    ///
+    /// - parameters:
+    ///     - requiresCleanClose: Determine if the client should throw when it is shutdown in a non-clean state
+    ///
+    /// - Note:
+    /// The `requiresCleanClose` will let the client do additional checks about its internal consistency on shutdown and
+    /// throw the appropriate error if needed. For instance, if its internal connection pool has any non-released connections,
+    /// this indicate shutdown was called too early before tasks were completed or explicitly canceled.
+    /// In general, setting this parameter to `true` should make it easier and faster to catch related programming errors.
+    public func syncShutdown(requiresCleanClose: Bool) throws {
```

Comment:
do we need this as public API? I think this can remain internal, no?

Reply:
@weissi  I think it might need to be public because `syncShutdown()` calls `syncShutdown(requiresCleanClose: false)` for backwards compatibility, but since users should preferably do a clean close, they need to be able to call `syncShutdown(requiresCleanClose: true)`. WDYT?

Reply:
but if you were a user of this library, wouldn't you be super confused on which shutdown to use and what parameter to pass? I'd just not expose `syncShutdown(requiresCleanClose:)` (can be used in tests of course). If we feel that it's super important, we can always make it public later

Reply:
I was worried about the user shutting down the client too early and having difficulties understanding why all their tasks are getting `cancelled`, but I recognise `requiresCleanClose` is a bit confusing and the docs I added are not as much clear as they could be. I'll remove it for now then, never too late to add it later indeed.

Reply:
Were you planning to remove this?